actsims/batchTools.py

```diff
@@ -0,0 +1,185 @@
+data_dir        = os.path.join(os.path.dirname(os.path.abspath(__file__)), '../data')
+
+
+class MR3PATCH_HELPER(object):
+    # a helper class to quickly generate sims for given patch
+    def __init__(self, set_idx, sim_idx, patch, cmb_type='LensedCMB', noisediag_only=False, dobeam=True, add_foregrounds=True, apply_window=True):
+        self.patch    = patch.lower()
+        assert(self.patch in ['boss', 'deep1', 'deep56', 'deep5', 'deep6', 'deep8'])
+        assert(cmb_type in ['LensedCMB', 'UnlensedCMB', 'LensedUnabberatedCMB'])
+
+        self.set_idx    = set_idx
+        self.sim_idx    = sim_idx
+        self.cmb_type   = cmb_type
+        self.alms_base  = {}
+        self.alms_patch = {}
+        self.signals    = {}
+        self.noises     = {}
+        self.psas       = []
+        self.psafs      = []
+        freqs           = [] 
+        self.template   = None
+        self.dobeam     = dobeam
+        for key in signal_dict['beamNames'].keys():
+            if self.patch in key: 
+                self.psafs.append(key)
+                self.psas.append(key[:-5])
+                freqs.append(key.split('_')[-1])
+            else: pass
+        self.psas  = list(set(self.psas))
+        self.psas.sort()
+        self.psafs.sort()
+        self.freqs = list(set(freqs))
+        self.freqs.sort()
+        for psaf in self.psafs:
+            self.noises[psaf] = {}
+        self.noisediag_only  = noisediag_only
+        self.add_foregrounds = add_foregrounds
+        self.apply_window    = apply_window
+
+    def get_signal_sim(self, psaf, save_alm=False, save_map=False):
+        assert(psaf in self.psafs)
+        freq = psaf.split('_')[-1]
+        print "loading sims for {}".format(psaf)
+        if self.signals.has_key(psaf): 
+            print "loading precomputed sim {}".format(psaf)
+            return self.signals[psaf].copy()
+        
+        alm_patch = None
+        if self.alms_patch.has_key(psaf):  
+            print "loading precomputed alm {}".format(psaf)
+            return self.signals[psaf].copy()
+            alm_patch = self.alms_patch[psaf].copy()
+        else:
+            if len(self.alms_base) == 0: self.load_alms_base()
+            alm_patch = self.alms_base[freq].copy()
+            if self.dobeam:
+                print "apply beam for alm {}".format(psaf)
+                beam_file = os.path.join(os.path.dirname(os.path.abspath(__file__)),signal_dict['beamNames'][psaf])
+                beam_data = (np.loadtxt(beam_file))[:,1]
+                for idx in range(alm_patch.shape[0]):
+                    alm_patch[idx] = hp.sphtfunc.almxfl(alm_patch[idx].copy(), beam_data)
+            else: pass
+            if save_alm: self.alms_patch[psaf] = alm_patch.copy()
+
+        signal = self.get_template()
+        curvedsky.alm2map(alm_patch, signal, spin = [0,2], verbose=True)
+        
+        if self.apply_window:
```

Comment:
I am trying to figure out how much of this dupicates what is in simTools.  Best not to have it in two places.  I am fine with scrapping simTools.py if we can get all its functionality with this class-based approach.  I think it means we have to change the interfacing with it.  

I had written actsims.simTools as a single function as an entryway  simTools.getActpolSim().  It seems like with this approach you have obviated the need for that -- maybe we should just make that change across the board?  To my knowledge the only people using the actsims code are me, DW, Mat and Steve.  What do you think of changing the interface?

Reply:
I think if we combine this module and Mat's new noise code, we should recover all the functionalities in simTools.py. But the current implementation isn't a good enough for the general release. As it is, you have to reinitialize the object for different sets and iteration number. Of course, that's something we can fix with more complicated bookkeeping inside the code. 

And if we are thinking of retiring simTools.py, we should start by raising depreciation warnings and phase it out in a couple iteration.

Reply:
I see, ok